Modules/_zstd/_zstdmodule.c

```diff
@@ -0,0 +1,968 @@
+    }
+
+    /* Allocate dict buffer */
+    dst_dict_bytes = PyBytes_FromStringAndSize(NULL, dict_size);
+    if (dst_dict_bytes == NULL) {
+        goto error;
+    }
+
+    /* Train the dictionary */
+    Py_BEGIN_ALLOW_THREADS
+    zstd_ret = ZDICT_trainFromBuffer(PyBytes_AS_STRING(dst_dict_bytes), dict_size,
+                                     PyBytes_AS_STRING(samples_bytes),
+                                     chunk_sizes, (uint32_t)chunks_number);
+    Py_END_ALLOW_THREADS
+
+    /* Check zstd dict error */
+    if (ZDICT_isError(zstd_ret)) {
+        _zstd_state* const _module_state = PyModule_GetState(module);
+        if (_module_state != NULL) {
+            set_zstd_error(_module_state, ERR_TRAIN_DICT, zstd_ret);
+        }
+        goto error;
+    }
+
+    /* Resize dict_buffer */
+    if (_PyBytes_Resize(&dst_dict_bytes, zstd_ret) < 0) {
+        goto error;
+    }
+
+    goto success;
+
+error:
+    Py_CLEAR(dst_dict_bytes);
+
+success:
+    PyMem_Free(chunk_sizes);
+    return dst_dict_bytes;
+}
+
+/*[clinic input]
+_zstd._finalize_dict
+
+        custom_dict_bytes: PyBytesObject
+            Custom dictionary content.
+        samples_bytes: PyBytesObject
+            Concatenation of samples.
+        samples_size_list: object(subclass_of='&PyList_Type')
+            List of samples' sizes.
+        dict_size: Py_ssize_t
+            The size of the dictionary.
+        compression_level: int
+            Optimize for a specific zstd compression level, 0 means default.
+        /
+
+Internal function, finalize a zstd dictionary.
+[clinic start generated code]*/
+
+static PyObject *
+_zstd__finalize_dict_impl(PyObject *module, PyBytesObject *custom_dict_bytes,
+                          PyBytesObject *samples_bytes,
+                          PyObject *samples_size_list, Py_ssize_t dict_size,
+                          int compression_level)
+/*[clinic end generated code: output=9c2a7d8c845cee93 input=ef6f4f2b0b7e178d]*/
+{
+    if (ZSTD_versionNumber() < 10405) {
+        /* Must be dynamically linked */
+        PyErr_Format(PyExc_NotImplementedError,
+                "The _finalize_dict function is only available when the underlying "
+                "zstd library's version is greater than or equal to v1.4.5. "
+                "The current zstd version is v%s.",
+                ZSTD_versionString());
+        return NULL;
+    }
+
+    Py_ssize_t chunks_number;
+    size_t *chunk_sizes = NULL;
+    PyObject *dst_dict_bytes = NULL;
+    size_t zstd_ret;
+    ZDICT_params_t params;
+    Py_ssize_t sizes_sum;
+    Py_ssize_t i;
+
+    /* Check arguments */
+    if (dict_size <= 0) {
+        PyErr_SetString(PyExc_ValueError, "dict_size argument should be positive number.");
+        return NULL;
+    }
+
+    chunks_number = Py_SIZE(samples_size_list);
+    if ((size_t) chunks_number > UINT32_MAX) {
+        PyErr_SetString(PyExc_ValueError,
+                        "The number of samples should <= UINT32_MAX.");
```

Comment:
Likewise, if this exception can be triggered by Python code, we should avoid using C types in it.

```suggestion
        PyErr_SetString(PyExc_ValueError,
                        "The number of samples must be less than 2**32.");
```

Reply:
We have other places (e.g., `_pysqlite_set_result`) where we mention INT_MAX but other places where we use `%i` and `INT_MAX`. I suggest using `%ju` to format `UINT32_MAX` (IIRC, it's `%u` for `uint32_t`).

Reply:
Changing existing usages might be tricky, but we should avoid introducing new ones.

Reply:
I actually myself introduced some when I implemented HMAC, so I'll amend this as well.

Reply:
I used a `%u` format with `UINT32_MAX` to present the range.